---
title: "BMI FTO Example PheWAS"
output: html_document
---
```{r setup, include=FALSE}
#knitr::opts_knit$set( root.dir = "~/")
```
## PheWAS R Package
This Rmd file will walk you through an example single SNP PheWAS. The data we will use today is simulated data, generated to reflect the real pattern of correlation found in the EHR. You have been provided one SNP, rs8050136, which is additively coded. There are three potential covariates included: age, is.male, and bmi. Remember that you can view documentation for the PheWAS package methods and other R functions using ?, e.g. ?phewas. In addition, the vignette contains a walkthrough and is available using vignette("PheWAS-package").

## Install PheWAS
We first need to install and load the v1.0 PheWAS package:
```{r, message=F}
if(!require(PheWAS)||packageVersion("PheWAS")<'0.99.4'){
  if(!require(devtools)){
    install.packages("devtools")
  }
  devtools::install_github("PheWAS/PheWAS")
  library(PheWAS)
}
if(!require(readr)) {
  install.packages("readr")
  library("readr")
} #For faster file loading
```

## Load data
There are a number of ways to load data, and we may have a variety of data sources for our analysis. First, we will load our genotype data from our plink raw file. It may be necessary to set or change your working directory (setwd). Alternatively, you can specify the complete path to the file you wish to load. Note that knitr requires us to use a special setup for the working directory above.
```{r}
#setwd("~/")
genotypes=read_delim("exome_chip_fto.raw",delim=" ") %>% transmute(id=FID,rs8050136_A)
```
We only need to keep the id and snp allele count columns. We can take a quick peek at our data to make sure it is coded properly and see our allele frequencies.
```{r}
table(genotypes$rs8050136_A)
```
Next, we read our billing code data from a csv file. This file includes the id of the individual, the vocabulary and code for this row (ICD9CM or ICD10CM), and the count of days that code occurred. We specify the data types of the columns as R may want to read ICD9CM codes as numeric values (which can lose trailing and leading zeros).
```{r}
icd_codes=read_csv("icd_codes.csv", col_types="ifci")
```
Lastly, we will read our covariates in. These are saved in an RData file, which is a compressed, R specific file format.
```{r}
load("covariates.RData")
```

## Prepare Data
Using the raw format data from plink means our genotype data will be ready to go. We are not going to transform our covariates either, so that leaves us to set up our phenotypes. The below call will use the default minimum code count of 2. This may take a few minutes to complete. Note that we are restricting some phenotypes that are specific to males or females, eg prostate cancer. We could also define an aggregate function here, as we are using count data and not dates of occurrence (the preferred input), but it will be handled by the default.
```{r}
phenotypes=createPhenotypes(icd_codes, 
                            id.sex = covariates %>% transmute(id, gender=ifelse(is.male,"M","F")))
```
We can actually export this data to a phenotype file that would work with plink if interested. These can be quite large, so we are restricting to a single phenotype here
```{r}
plinkPhewasExport(phenotypes %>% select(id, `250.2`),file="plink.pheno")
```
We can pass in these separate data frames we have now loaded directly into the phewas method, but we may want to investigate our data later, so we will merge it ourselves. Note that we need to specify the names of our id columns using the 'by' parameter.
```{r}
data=inner_join(inner_join(covariates, genotypes),phenotypes)
```

## Run PheWAS
Running the PheWAS on a single SNP will not take very long, but we can speed it up using more cores. We will pass phewas the data frame we have merged already, so we need to tell it the names of the columns that contain the phenotypes, genotypes, and covariates. I get the names from the "phenotype" data frame and remove "FID", which is the first element. Note that you can adjust the number of cores as applicable to your system- be sure to play nicely if you are sharing a server!
```{r}
results=phewas(phenotypes=names(phenotypes)[-1], genotypes=c("rs8050136_A"), covariates=c("age", "is.male"), data=data, cores=4)
```

## View Our Results
Checking out our results both as a table and figure can be helpful. There are many ways to order and view the results in R:
```{r}
head(addPhecodeInfo(results) %>% arrange(p)) 
```

Head gives the top n rows (6 by default), addPhecodeInfo adds the name of the phewas code to the data frame, and arrange orders the data frame according to p-value.
We can report all associations from the PheWAS with p<1e-6 using the code below. There's also a handy method for creating an easy to use table in your knitted document! One can export the table with write.csv(), enabling it to be read by any spreadsheet software.
```{r}
phewasDT(results)
write.csv(addPhecodeInfo(results) %>% arrange(p) %>% filter(p<1e-6) %>% select(phenotype, description, snp, OR, p, n_total, n_cases, n_controls), file="top-results.csv")
```

Note that our top results are three phecodes that look very similar. As you may remember, phecodes are hierarchical. We shouldn't be surprised to see the overlap in association here- the cases for each test are largely overlapped.

## Plot the Results
Creating a rough visualization is straight forward.
```{r}
phewasManhattan(results)
```

Try tweaking the plot using parameters, such as 'title'. You can view other options using ?phenotypePlot, ?phenotypeManhattan, and ?phewasManhattan (they all direct to the same place). Remember that most parameters from phenotypePlot can be used in the "higher level" plotting functions as well. 
```{r}
my_plot=phewasManhattan(results, title="My example plot", OR.direction=TRUE)
my_plot
```

If you wish to plot directly to file, you can use png or pdf:
```{r}
png("MyPlot.png")
my_plot
dev.off()
```

Note that changing the size and scale of the output plot can affect how it looks. The method ggsave from ggplot2 can be helpful for getting consistent output.

## Adjusting for BMI (sometimes)
Let's take a look at the effects of BMI on these associations. The phewas_ext function allows one run sets of associations with defined sets of covariates by providing a list of vectors instead of just one vector of covariates. Note that this will run the same analysis we did before, but now we will have a more complete result set.
```{r}
results=phewas_ext(phenotypes=names(phenotypes)[-1], genotypes=c("rs8050136_A"), covariates=list(c("age", "is.male"),c("age", "is.male","BMI")), data=data, cores=4)
```

## Checking our BMI-adjusted results
If we want to see what the results look like for a single group of covariates, we will need to filter our results.
```{r}
results_bmi_adjusted = results %>% filter(covariates=='age,is.male,BMI')
phewasManhattan(results_bmi_adjusted, title="Results adjusted for BMI",y.axis.interval=1)
```

## Phenotypes predicting phenotypes
What happens if we predict phenotypes with our bmi variable?
```{r }
results_bmi=phewas(phenotypes=names(phenotypes)[-1], predictors = c("BMI"), covariates=c("age", "is.male"), data=data, cores=4, additive.genotypes = F)
```
One will often see incredibly strong results in these cases, so don't be surprised! In addition, here's the raw code for creating a browsable data table- easily customized from the above default method phewasDT.
```{r}
DT::datatable(results_bmi %>% addPhecodeInfo() %>% arrange(p) %>% filter(!is.na(p)) %>% transmute(phenotype, description, predictor, OR=round(OR,2), p=signif(p,2), n_total, n_cases, n_controls))
```


## Conclusions
Congratulations, you have successfully run a PheWAS using R (and perhaps knitted your first markdown file)!
